src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php

```diff
@@ -42,15 +44,31 @@ protected function processValue($value, $isRoot = false)
 
         $arguments = $value->getArguments();
         if (!isset($arguments[0]) || !\is_array($arguments[0])) {
-            throw new InvalidArgumentException(sprintf('Invalid definition for service "%s": an array of references is expected as first argument when the "container.service_locator" tag is set.', $this->currentId));
+            throw new InvalidArgumentException(sprintf('Invalid definition for service "%s": an array of references or service types is expected as first argument when the "container.service_locator" tag is set.', $this->currentId));
         }
 
         foreach ($arguments[0] as $k => $v) {
             if ($v instanceof ServiceClosureArgument) {
                 continue;
             }
+
+            if (\is_string($type = $v) && preg_match('/^\??[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+(?:\\\\[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+)*+$/', $type)) {
```

Comment:
this will allow `<argument>Psr\Logger\LoggerInterface</argument>`
alongside with `<argument type="service" id="Psr\Logger\LoggerInterface" />`
I read too fast this part of your proposal. I think it's not a good idea. It would create an inconsistency in the way the configuration work.
But being able to omit the "key" part looks like a good idea to me.